### PR TITLE
[Lazy evaluation] Pytato Array Context with transformations

### PR DESCRIPTION
**Pull request by kaushikcfd**
Draft because:
- [x] Implement a global reduction parallelization strategy
- [x] https://github.com/inducer/loopy/pull/224
- [x] https://github.com/inducer/loopy/pull/476
- [x] https://github.com/inducer/arraycontext/pull/68
- [x] https://github.com/inducer/pytato/pull/127

Related/linked PRs:
- [x] https://github.com/inducer/pytato/pull/169
- [x] https://github.com/inducer/arraycontext/pull/104

### CODE REVIEW

File: meshmode/array_context.py [hunk 1]
```diff
@@ -309,4 +310,89 @@ def transform_loopy_program(self, t_unit):
         # FIXME: Do not parallelize for now.
         return t_unit
 
+
+@for_each_kernel
+def _single_grid_work_group_transform(kernel, cl_device):
```
Comment (kaushikcfd) [inline meshmode/array_context.py:315]: @inducer said in a personal meeting that this logic should also handle reductions kernels.

File: meshmode/array_context.py [hunk 2]
```diff
@@ -309,4 +310,89 @@ def transform_loopy_program(self, t_unit):
...
+            if len(insn.within_inames) == 1:
+                iname, = insn.within_inames
+
+                kernel = lp.split_iname(kernel, iname,
+                                        ngroups * l_zero_size * l_one_size)
+                kernel = lp.split_iname(kernel, f"{iname}_inner",
+                                        l_zero_size, inner_tag="l.0")
+                kernel = lp.split_iname(kernel, f"{iname}_inner_outer",
+                                        l_one_size, inner_tag="l.1",
+                                        outer_tag="g.0")
```
Comment (inducer) [inline meshmode/array_context.py:344]: Doesn't this assume that each iname is only used by exactly one statement?

File: meshmode/array_context.py [hunk 3]
```diff
@@ -309,4 +310,89 @@ def transform_loopy_program(self, t_unit):
...
+            # must be a callable kernel, don't touch.
+            pass
+        elif isinstance(insn, lp.Assignment):
+            bigger_loop = None
+            smaller_loop = None
+
+            if len(insn.within_inames) == 0:
+                continue
+
+            if len(insn.within_inames) == 1:
```
Comment (inducer) [inline meshmode/array_context.py:335]: * Which type of kernels are those?
* I would like it if this were more guided by metadata, along the lines of what the pyopencl actx does.

**Review by inducer — COMMENTED** [review]
Thanks! A few quick drive-by thoughts below.

File: meshmode/array_context.py [hunk 4]
```diff
@@ -392,6 +479,9 @@ def transform_loopy_program(self, t_unit):
         import loopy as lp
         t_unit = _single_grid_work_group_transform(t_unit, self.queue.device)
         t_unit = lp.set_options(t_unit, "insert_gbarriers")
+        t_unit = lp.linearize(lp.preprocess_kernel(t_unit))
+        t_unit = _alias_global_temporaries(t_unit)
```
Comment (inducer) [inline meshmode/array_context.py:483]: My recollection of our discussion was that we'd do this without aliasing... am I remembering wrong? If not, what made you change your mind?

Comment (kaushikcfd) [reply meshmode/array_context.py:483]: I thought we wouldn't alias at the `pytato`'s CodeGenMapper stage, but post-linearization grabbing hold of dead-temporaries would be trivial. I.e. I had thought we would alias the global temporaries not at the pytato-level but downstream as loopy transformation.

Comment (kaushikcfd) [reply meshmode/array_context.py:483]: x-ref: https://github.com/inducer/pytato/issues/122

**Comment by matthiasdiener** [comment]
When trying this on the mirgecom examples, I'm getting this type of error:

```
  File "/Users/mdiener/Work/emirge/grudge/grudge/discretization.py", line 623, in nodes
    return self.discr_from_dd(dd).nodes()
  File "/Users/mdiener/Work/emirge/meshmode/meshmode/discretization/__init__.py", line 610, in nodes
    result = make_obj_array([
  File "/Users/mdiener/Work/emirge/meshmode/meshmode/discretization/__init__.py", line 611, in <listcomp>
    _DOFArray(None, tuple([
  File "/Users/mdiener/Work/emirge/meshmode/meshmode/discretization/__init__.py", line 612, in <listcomp>
    actx.freeze(resample_mesh_nodes(grp, iaxis)) for grp in self.groups
  File "/Users/mdiener/Work/emirge/arraycontext/arraycontext/impl/pytato/__init__.py", line 133, in freeze
    pt_prg = pt_prg.with_transformed_program(self.transform_loopy_program)
  File "/Users/mdiener/Work/emirge/pytato/pytato/target/loopy/__init__.py", line 124, in with_transformed_program
    return self.copy(program=f(self.program))
  File "/Users/mdiener/Work/emirge/meshmode/meshmode/array_context.py", line 482, in transform_loopy_program
    t_unit = lp.linearize(lp.preprocess_kernel(t_unit))
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 2201, in linearize
    knl = get_one_linearized_kernel(knl,
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 2178, in get_one_linearized_kernel
    result = _get_one_scheduled_kernel_inner(kernel,
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 2146, in _get_one_scheduled_kernel_inner
    return next(iter(generate_loop_schedules(kernel, callables_table)))
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 1953, in generate_loop_schedules
    yield from generate_loop_schedules_inner(kernel,
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 2084, in generate_loop_schedules_inner
    gen_sched = insert_barriers(kernel, gen_sched,
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 1919, in insert_barriers
    result = insert_barriers_at_outer_level(result)
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 1838, in insert_barriers_at_outer_level
    append_barrier_or_raise_error(
  File "/Users/mdiener/Work/emirge/loopy/loopy/schedule/__init__.py", line 1760, in append_barrier_or_raise_error
    raise MissingBarrierError(
loopy.diagnostic.MissingBarrierError: _pt_kernel: Dependency '_pt_out_store depends on call_lp_nodes' (for variable '_pt_temp') requires synchronization by a global barrier (add a 'no_sync_with' instruction option to state that no synchronization is needed)
```

**Comment by kaushikcfd** [comment]
@matthiasdiener: Sorry, the transformations in this branch depend on some PRs on the pytato/loopy/arraycontext end. I've edited the description to record that.

File: meshmode/array_context.py (hunk 3 above)
Comment (kaushikcfd) [reply meshmode/array_context.py:359]: We could borrow the unifier from `FusionContractorArrayContext`, but that would add a dependency on https://github.com/inducer/pytato/pull/224. If that goes in before this PR, I will put that here.

**Comment by kaushikcfd** [comment]
Closing this as https://github.com/inducer/arraycontext/pull/216 provides a cleaner implementation.

**Pull request closed without merging**